From 1b9b325d9feda98b7dd0f3897390b591de0b82e9 Mon Sep 17 00:00:00 2001
From: yliu <yliu@apache.org>
Date: Tue, 31 Mar 2015 21:16:13 +0800
Subject: [PATCH 0829/1023] HDFS-7997. The first non-existing xattr should
 also throw IOException. (zhouyingchao via yliu)

(cherry picked from commit e4cd67e9b0bcc4e5aedb8c189dc780326c69b032)

Change-Id: Iece0f1cc7986c2626d716e638d886ff746720cab
(cherry picked from commit a7d07f34ede2fe3ef79a2ff92b7a92ee01e3c449)
---
 .../hadoop/hdfs/server/namenode/FSNamesystem.java  |    3 ++-
 .../src/test/resources/testXAttrConf.xml           |    4 ++--
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
index 84fd5ec..0b2d0e2 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
@@ -9257,7 +9257,8 @@ private void checkXAttrSize(XAttr xAttr) {
         return filteredAll;
       } else {
         if (filteredAll == null || filteredAll.isEmpty()) {
-          return null;
+          throw new IOException(
+              "At least one of the attributes provided was not found.");
         }
         List<XAttr> toGet = Lists.newArrayListWithCapacity(xAttrs.size());
         for (XAttr xAttr : xAttrs) {
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testXAttrConf.xml b/hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testXAttrConf.xml
index 9c66cba..c2e836c 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testXAttrConf.xml
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testXAttrConf.xml
@@ -420,8 +420,8 @@
       </cleanup-commands>
       <comparators>
       	<comparator>
-          <type>ExactComparator</type>
-          <expected-output># file: /file1#LF#</expected-output>
+          <type>SubstringComparator</type>
+          <expected-output>At least one of the attributes provided was not found</expected-output>
         </comparator>
       </comparators>
     </test>
-- 
1.7.9.5

